### PR TITLE
Add metrics for deployment, replicaset, replication_controller, statefulset and hpa

### PR DESCRIPTION
**Pull request by ChrsMark**
## Changes

Part of https://github.com/open-telemetry/semantic-conventions/issues/1032.

Fixes https://github.com/open-telemetry/semantic-conventions/issues/1644
Fixes https://github.com/open-telemetry/semantic-conventions/issues/1637

Adds the following metrics:

- [`k8s.deployment.desired_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sdeploymentdesired) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/deployment/deployments.go#L32)) [⚠️ BREAKING change for the Collector]
- [`k8s.deployment.available_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sdeploymentavailable) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/deployment/deployments.go#L33)) [⚠️ BREAKING change for the Collector]
- [`k8s.replicaset.desired_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplicasetdesired) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicaset/replicasets.go#L31)) [⚠️ BREAKING change for the Collector]
- [`k8s.replicaset.available_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplicasetavailable) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicaset/replicasets.go#L32)) [⚠️ BREAKING change for the Collector]
- [`k8s.replication_controller.desired_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplication_controllerdesired) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicationcontroller/replicationcontrollers.go#L17-L18)) [⚠️ BREAKING change for the Collector]
- [`k8s.replication_controller.available_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplication_controlleravailable) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicationcontroller/replicationcontrollers.go#L17-L18)) [⚠️ BREAKING change for the Collector]
- [`k8s.statefulset.desired_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetdesired_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [`k8s.statefulset.ready_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetready_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [`k8s.statefulset.current_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetcurrent_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [`k8s.statefulset.updated_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetupdated_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [`k8s.hpa.desired_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpadesired_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18))  [⚠️ BREAKING change for the Collector]
- [`k8s.hpa.current_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpacurrent_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18))  [⚠️ BREAKING change for the Collector]
- [`k8s.hpa.max_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpamax_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18))  [⚠️ BREAKING change for the Collector]
- [`k8s.hpa.min_pods`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpamin_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18))  [⚠️ BREAKING change for the Collector]

These metrics have been in use by the Opentelemetry Collector and specifically the [`k8scluster` receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md) component.

Breaking changes between the Collector and the Semconv will be listed in the k8s migration guide once https://github.com/open-telemetry/semantic-conventions/pull/1636 is merged.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

**Comment by ChrsMark** [comment]
@open-telemetry/specs-semconv-approvers should the `.count` suffix be applied here based on the guideline from https://github.com/open-telemetry/semantic-conventions/blob/2d34907ac6d08735060f19c759771dd37ff36ee7/docs/general/metrics.md?plain=1#L148?

We don't use a pluralized metric name here but I want to double check if this rule should be taken into account.

It would be preferable if we define this metrics as is to minimize breaking changes of the Collector.

File: model/k8s/metrics.yaml
Comment (tetianakravchenko) [inline model/k8s/metrics.yaml:169]: shouldn't it be `availableReplicas` instead - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/k8sclusterreceiver/internal/replicaset/replicasets.go#L24?
```suggestion
      This metric aligns with the `availableReplicas` field of the
```

File: model/k8s/metrics.yaml
Comment (tetianakravchenko) [inline model/k8s/metrics.yaml:147]:
```suggestion
      This metric aligns with the `availableReplicas` field of the
```
based on https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/k8sclusterreceiver/internal/deployment/deployments.go#L26

File: model/k8s/metrics.yaml
Comment (ChrsMark) [reply model/k8s/metrics.yaml:169]: Correct!

File: docs/system/k8s-metrics.md
Comment (tetianakravchenko) [inline docs/system/k8s-metrics.md:409]: readme file needs to be updated accordingly

**Review by tetianakravchenko — APPROVED** [review]

**Comment by ChrsMark** [comment]
We need to have this aligned with https://github.com/open-telemetry/semantic-conventions/issues/1637 and https://github.com/open-telemetry/semantic-conventions/issues/1644.

**Review by jaronoff97 — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
Taking back my approval until discussion in https://github.com/open-telemetry/semantic-conventions/issues/1637#issuecomment-2515544158 is resolved

**Comment by ChrsMark** [comment]
I have updated the PR to include the `statefulset` and `hpa` metrics since those are related to the consistency discussion (https://github.com/open-telemetry/semantic-conventions/issues/1637#issuecomment-2515544158).

If want to achieve consistency while also being semantically correct we either need to use the `_pods` or the `_replicas` suffix. I chose to use the `_pods` suffix as more "user friendly".

Providing bellow the current state (copied from the PR's description):

- [k8s.deployment.desired_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sdeploymentdesired) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/deployment/deployments.go#L32)) [⚠️ BREAKING change for the Collector]
- [k8s.deployment.available_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sdeploymentavailable) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/deployment/deployments.go#L33)) [⚠️ BREAKING change for the Collector]
- [k8s.replicaset.desired_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplicasetdesired) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicaset/replicasets.go#L31)) [⚠️ BREAKING change for the Collector]
- [k8s.replicaset.available_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplicasetavailable) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicaset/replicasets.go#L32)) [⚠️ BREAKING change for the Collector]
- [k8s.replication_controller.desired_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplication_controllerdesired) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicationcontroller/replicationcontrollers.go#L17-L18)) [⚠️ BREAKING change for the Collector]
- [k8s.replication_controller.available_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sreplication_controlleravailable) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/replicationcontroller/replicationcontrollers.go#L17-L18)) [⚠️ BREAKING change for the Collector]
- [k8s.statefulset.desired_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetdesired_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [k8s.statefulset.ready_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetready_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [k8s.statefulset.current_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetcurrent_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [k8s.statefulset.updated_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8sstatefulsetupdated_pods) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9dfa2f7813b11500d001622f3d8c1dd8d9ac58fd/receiver/k8sclusterreceiver/internal/statefulset/statefulsets.go#L42-L45)) [ ✅ not a breaking change for the Collector]
- [k8s.hpa.desired_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpadesired_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18)) [⚠️ BREAKING change for the Collector]
- [k8s.hpa.current_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpacurrent_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18)) [⚠️ BREAKING change for the Collector]
- [k8s.hpa.max_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpamax_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18)) [⚠️ BREAKING change for the Collector]
- [k8s.hpa.min_pods](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.114.0/receiver/k8sclusterreceiver/documentation.md#k8shpamin_replicas) (emitted by [collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/06dd7e27fc5bb6dc151086bdb384297bc1d777c8/receiver/k8sclusterreceiver/internal/hpa/hpa.go#L15-L18)) [⚠️ BREAKING change for the Collector]

File: model/k8s/metrics.yaml
Comment (tetianakravchenko) [inline model/k8s/metrics.yaml:243]:
```suggestion
    brief: "Desired number of replicas of pods managed by this horizontal pod autoscaler, as last calculated by the autoscaler"
```
?

File: model/k8s/metrics.yaml
Comment (tetianakravchenko) [inline model/k8s/metrics.yaml:253]:
```suggestion
    brief: "Current number of replicas of pods managed by this horizontal pod autoscaler, as last seen by the autoscaler"
```

File: model/k8s/metrics.yaml
Comment (tetianakravchenko) [inline model/k8s/metrics.yaml:265]:
```suggestion
      This metric aligns with the `maxReplicas` field of the
```

File: model/k8s/metrics.yaml
Comment (tetianakravchenko) [inline model/k8s/metrics.yaml:275]:
```suggestion
      This metric aligns with the `minReplicas` field of the
```

**Review by tetianakravchenko — APPROVED** [review]

**Comment by ChrsMark** [comment]
Listed the Collector's breaking changes in [b469b4e](https://github.com/open-telemetry/semantic-conventions/pull/1636/commits/b469b4efd92806ff83b8f4a6d5d5f60c0308e4d6).

@dashpole @dmitryax @jaronoff97 feel free to take another look once you get the chance.

File: .chloggen/add_deploy_replica_metrics.yaml
Comment (jaronoff97) [inline .chloggen/add_deploy_replica_metrics.yaml:22]: it may be worth writing in the subtext here why this change is being made. Calling out that this is being done so we align more closely with the recommendations from Kubernetes.

**Review by jaronoff97 — APPROVED** [review]

**Review by jaronoff97 — APPROVED** [review]

File: docs/system/k8s-metrics.md
Comment (jinja2) [inline docs/system/k8s-metrics.md:397]: Should we explicitly clarify that the `this deployment` refers to the specific deployment resource this metric is for?

Comment (jinja2) [reply docs/system/k8s-metrics.md:397]: I see there is a note earlier in this doc which states to refer to the relevant k8s resource.

Comment (ChrsMark) [reply docs/system/k8s-metrics.md:397]: Yeap, probably we will need to iterate on this to improve the note: https://github.com/open-telemetry/semantic-conventions/pull/1649#discussion_r1872258043

**Comment by jinja2** [comment]
We have a few minor discrepancies in resource/metric names. While I don't think we need to change any existing names, I would like to note them down for future reference.

1. Short/Full names for k8s type - HPA seems to be the only resource type that uses the short name. The full object kind can be a mouthful, but best to stick with the full names as long as they are a reasonable length.
2. k8s object kinds use PascalCase. Some resources mirror this with underscores in the attr/metric names (e.g. `k8s.replication_controller` for k8s obj `ReplicationController`), while others don't (e.g. `k8s.statefulset` for k8s obj `StatefulSet`). I think we should be consistent with the underscore convention.

**Comment by jinja2** [comment]
I think [this](https://github.com/open-telemetry/semantic-conventions/pull/1660#issuecomment-2532069260) observation from @dashpole might be relevant to some of the metrics here. Most of these timeseries seem meaningfully additive

**Comment by ChrsMark** [comment]
> I think [this](https://github.com/open-telemetry/semantic-conventions/pull/1660#issuecomment-2532069260) observation from @dashpole might be relevant to some of the metrics here. Most of these timeseries seem meaningfully additive

As discussed in the K8s SemConv WG meeting today, we agree on making the change to `updowncounter` for these metrics (I think this will also apply to https://github.com/open-telemetry/semantic-conventions/pull/1649). I will update the PR(s) to reflect this.

**Review by dashpole — APPROVED** [review]

**Review by jaronoff97 — APPROVED** [review]
I appreciate that we moved from repliaca -> pods for the convention here. I think that change will make the user experience a bit simpler as well. Thank you!

**Comment by ChrsMark** [comment]
@open-telemetry/specs-semconv-maintainers could you also take a look please?

note: I'm [suggesting](https://github.com/open-telemetry/semantic-conventions/pull/1636#discussion_r1880207376) we introduce the metrics as-is and follow-up for enhancing their descriptions more to cover their specific relations with resources.

**Review by lmolkova — APPROVED** [review]

**Comment by ChrsMark** [comment]
I added resource correlation note per metric with [1f91666](https://github.com/open-telemetry/semantic-conventions/pull/1636/commits/1f91666efc43454add1628f6b49da43efb7dd6b4) according to the discussion at https://github.com/open-telemetry/semantic-conventions/pull/1649.

**Review by jsuereth — APPROVED** [review]

File: .chloggen/add_deploy_replica_metrics.yaml
Comment (ChrsMark) [reply .chloggen/add_deploy_replica_metrics.yaml]: @jaronoff97 this should be good to get resolved, right? This is the only blocking the merge now :).

Comment (jaronoff97) [reply .chloggen/add_deploy_replica_metrics.yaml]: yep! thank you.

**Pull request merged**
